src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Cache.cs

```diff
@@ -453,40 +453,83 @@ public static ulong GetKey(ReadOnlySpan<byte> name)
             ref byte reference = ref MemoryMarshal.GetReference(name);
             int length = name.Length;
 
-            if (length > 7)
+            if (BitConverter.IsLittleEndian)
             {
-                key = Unsafe.ReadUnaligned<ulong>(ref reference) & 0x00ffffffffffffffL;
-                key |= (ulong)Math.Min(length, 0xff) << 56;
+                if (length > 7)
+                {
+                    key = Unsafe.ReadUnaligned<ulong>(ref reference) & 0x00ffffffffffffffL;
+                    key |= (ulong)Math.Min(length, 0xff) << 56;
+                }
+                else
+                {
+                    key =
+                        length > 5 ? Unsafe.ReadUnaligned<uint>(ref reference) | (ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 32 :
+                        length > 3 ? Unsafe.ReadUnaligned<uint>(ref reference) :
+                        length > 1 ? Unsafe.ReadUnaligned<ushort>(ref reference) : 0UL;
+                    key |= (ulong)length << 56;
+
+                    if ((length & 1) != 0)
+                    {
+                        var offset = length - 1;
+                        key |= (ulong)Unsafe.Add(ref reference, offset) << (offset * 8);
+                    }
+                }
             }
             else
             {
-                key =
-                    length > 5 ? Unsafe.ReadUnaligned<uint>(ref reference) | (ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 32 :
-                    length > 3 ? Unsafe.ReadUnaligned<uint>(ref reference) :
-                    length > 1 ? Unsafe.ReadUnaligned<ushort>(ref reference) : 0UL;
-                key |= (ulong)length << 56;
-
-                if ((length & 1) != 0)
+                if (length > 7)
+                {
+                    key = Unsafe.ReadUnaligned<ulong>(ref reference) & 0xffffffffffffff00L;
+                    key |= (ulong)(uint)Math.Min(length, 0xff);
+                }
+                else
                 {
-                    var offset = length - 1;
-                    key |= (ulong)Unsafe.Add(ref reference, offset) << (offset * 8);
+                    key =
+                        length > 5 ? ((ulong)Unsafe.ReadUnaligned<uint>(ref reference) << 32) | ((ulong)Unsafe.ReadUnaligned<ushort>(ref Unsafe.Add(ref reference, 4)) << 16) :
+                        length > 3 ? ((ulong)Unsafe.ReadUnaligned<uint>(ref reference) << 32) :
+                        length > 1 ? ((ulong)Unsafe.ReadUnaligned<ushort>(ref reference) << 48) : 0UL;
+                    key |= (ulong)(uint)length;
+
+                    if ((length & 1) != 0)
+                    {
+                        var offset = length - 1;
+                        key |= (ulong)Unsafe.Add(ref reference, offset) << ((7 - offset) * 8);
+                    }
                 }
             }
```

Comment:
Does this actually need to be differentiated based on endianness?  What breaks if the key itself just ends up being different on big-endian boxes?

Reply:
@stephentoub, the keys are still different after this change.   Making them exactly identical would be possible but slower.  With this proposed patch, they are computed with the same efficiency, and they now at least have the same general properties on little-endian as on big-endian.

The current code without the patch does not maintain those same properties. For example, if length > 7 the current code would not actually hash the first 7 bytes (0 .. 6) but skip the first byte and hash the next 7 (1 .. 7).    I don't think this would be expected here.

In any case, the current code fails its assertion on big-endian, so some changes are needed in any case.

Reply:
What is the actual impact of that, other than failed asserts (which can be fixed)?  If you leave the code as is (but fix the asserts), does something break?  Does something run measurably slower?

Reply:
I don't think anything breaks except for the assert.  I have not really run any benchmarks yet, so I don't have any hard facts about performance.

However, it's pretty much impossible to "fix" the assert if we don't change the code, since the properties it wants to verify simply aren't true - there is no fixed place in the key where any of the input bytes will reliably end up, it's completely different depending on the initial length.   The existing code simply makes too many fundamentally little-endian assumptions.

Reply:
Let's just only do the assert on little endian then.

Reply:
OK, fine with me.  Patch updated.